qiskit_metal/analyses/quantization/lom_core_analysis.py

```diff
@@ -0,0 +1,1267 @@
+    # is the original node that is replaced by the ith junction; And an array of these arrays can
+    # represent all possible mappings
+
+    all_combos = _product_no_duplicate(*tuple(junctions.values()))
+    all_maps = [dict(zip(x, j_list)) for x in all_combos]
+
+    _transforms = []
+    for _map in all_maps:
+        new_basis = []
+        num_negative_nodes = 0
+        for n in orig_nodes:
+            if n not in _map:
+                new_basis.append(n)
+            else:
+                new_basis.append(_map[n])
+                if junctions[_map[n]][1] == n:  # "negative" node
+                    num_negative_nodes += 1
+
+        _transforms.append(
+            BasisTransform(orig_nodes, new_basis, num_negative_nodes))
+
+    if choose_least_num_neg:
+        _transforms.sort(key=lambda t: t.num_negative_nodes)
+        return _transforms[0]
+
+    return _transforms
+
+
+class LabeledNdarray(np.ndarray):
+    """np.ndarray subclass that shows labels when the array is printed.
+
+    The constructor only accepts a square matrix. Sliced LabeledNdarray will still be
+    a LabeledNdarray but the labels are not preserved: if the resulting matrix is still
+    a square matrix, the labels will just be [0, 1, ...<array_row_number>-1]; and if the
+    the resulting marix is not a square matrix, there will be no labels.
+
+    """
+
+    def __new__(cls, input_array: np.ndarray, labels=None):
+        if input_array.shape[0] != input_array.shape[1]:
+            raise ValueError(
+                'LabeledNdarray constructor only supports square matrices.')
+        obj = np.asarray(input_array).view(cls)
+        obj.labels = labels if labels is not None else [
+            str(x) for x in range(obj.shape[0])
+        ]
+        return obj
+
+    def to_dataframe(self):
+        if len(self.shape) == 1 or self.shape[0] != self.shape[1]:
+            raise ValueError(
+                'Only square matrices are labeled and can be exported to dataframe.'
+            )
+        return _make_cmat_df(self, self.labels)
+
+    def __array_finalize__(self, obj):
+        if obj is None:
+            return
+        if self.shape != obj.shape:
+            self.labels = [str(x) for x in range(self.shape[0])]
+        else:
+            self.labels = getattr(obj, 'labels',
+                                  [str(x) for x in range(self.shape[0])])
+
+    def __repr__(self):
+        if len(self.shape) == 1 or self.shape[0] != self.shape[1]:
+            return f'{self.view(np.ndarray)}'
+        return f'{_make_cmat_df(self, self.labels)}'
+
+    def __str__(self):
+        if len(self.shape) == 1 or self.shape[0] != self.shape[1]:
+            return f'{self.view(np.ndarray)}'
+        return f'{_make_cmat_df(self, self.labels)}'
+
+
+#----------------------------------------------------------------------------------------------------
+
+
+def _maybe_remove_grd_node_then_cache(f):
```

Comment:
lol?

Reply:
it's "maybe" because the function might not remove grd_node if the parameter "ignore ground node" is set false. "Maybe" in function names for conditional behaviors like this are common practices, IMO.